File: projects/P/python/P.py
# !/usr/bin/env python
# coding=utf-8
__author__ = 'Jayin Ton'

import re

pattern = re.compile(r'\{\s*\w+\s*\}')


class P(object):
    def __init__(self, string):
        self._str = string
        self._kv = {}

    @classmethod
    def _from(cls, string):
        return cls(string)

    def put(self, key, value):
        self._kv[key] = value
        return self

    def format(self):
        self._str = pattern.sub(self.__opt, self._str)
        return self._str

    def __opt(self, matchobj):
        return self._kv.get(matchobj.group(0).replace('{', '') \
                            .replace('}', '').strip(), matchobj.group(0))

File: projects/P/ruby/P.rb
class P

    def initialize(string)
        @kv = {}
        @str = string
    end

    def P.from(string)
        P.new string
    end

    def put(key,value)
        @kv[key] = value
        self
    end

    def format()
        @kv.each do |key,value|
            @str.gsub!(/{\s*\w+\s*}/) { |match| 
                match.gsub!(/{/,'').gsub!(/}/,'').strip!
                @kv[match]
            }
        end
        @str
    end
end


